### PR TITLE
customize: Add --network-nmstate option

### PR DESCRIPTION
**Pull request by qinqon**
It's possible to generate NetworkManager keyfiles from a Nmstate
configuration. This change add a new option to use Nmstate format
instead of multiple NetworkManager keyfiles to specify network
configuration.

The output from nmstate `gen_conf` is a HashMap of backend name to nmconnection vectors with the following format 

```yaml
NetworkManager:
  - "[connection]\nautoconnect=true\nautoconnect-slaves=-1\nid=dummyfoo\ninterface-name=dummyfoo\ntype=dummy\nuuid=157d8614-c3bd-4a10-9330-310fcf692468\n\n[ipv4]\nmethod=disabled\n\n[ipv6]\nmethod=disabled\n"
```

One call to `gen_conf` can dump multiple keyfiles from the "NetworkManager" backend.

### CODE REVIEW

File: src/cmdline/mod.rs
Comment (jlebon) [inline src/cmdline/mod.rs:292]: For consistency with `--network-keyfile`, let's make this be a `Vec<String>` too instead?

File: src/cmdline/mod.rs
Comment (jlebon) [inline src/cmdline/mod.rs:284]: It looks like it's typed "NMState".

File: src/live/customize.rs
Comment (jlebon) [inline src/live/customize.rs:184]: These two `match`es look like they could be collapsed to two `.context(...)?` instead to handle errors. That'll help unindent things.

File: Cargo.toml
Comment (jlebon) [inline Cargo.toml:75]: This looks like leftover from testing.

File: fixtures/customize/installer-test.json
Comment (jlebon) [inline fixtures/customize/installer-test.json:1]: Aren't NMState files YAML? Or does it support both YAML and JSON? (Or is this actually JSON being parsed as YAML?)

Would be good to add a YAML fixture too/instead.

**Review by jlebon — COMMENTED** [review]
Cool, thanks! Assuming getting the deps right in Fedora isn't too much work, this SGTM. I think even if we end up doing https://github.com/coreos/fedora-coreos-tracker/issues/1175, this still has value because it allows configuring the network in the initramfs of the live system too using NMState files.

File: src/live/customize.rs
Comment (qinqon) [reply src/live/customize.rs:184]: Tried that first but the NmstateError is not implemented the expected trait.

File: Cargo.toml
Comment (qinqon) [reply Cargo.toml:75]: The nmstate rust lib is not published yet at cargo.io they are going to do so during this week I think. Meanwhile I point it there to my local clone, this is draft PR.

File: src/cmdline/mod.rs
Comment (qinqon) [reply src/cmdline/mod.rs:284]: Official name is with just capital N, https://github.com/nmstate/nmstate/pull/224.

Comment (jlebon) [reply src/cmdline/mod.rs:284]: Ahh interesting. Looks like the official website needs to be fixed then: https://nmstate.io/.

File: tests/images/customize.sh
Comment (bgilbert) [inline tests/images/customize.sh:61]: This is not true currently, so you'll need to update the `live.ign` and `dest.ign` fixtures.

File: src/cmdline/mod.rs
Comment (bgilbert) [reply src/cmdline/mod.rs:284]: Once CI is working for this PR, it'll reject this line as being too long.  You can check manually with `tests/help.sh`.

File: src/cmdline/mod.rs
Comment (bgilbert) [inline src/cmdline/mod.rs:286]: Maybe let's say something like "using a NetworkManager keyfile generated from the specified Nmstate config"?

File: src/cmdline/mod.rs
Comment (bgilbert) [inline src/cmdline/mod.rs:293]: Nit: stray blank line.

File: src/live/customize.rs
Comment (bgilbert) [inline src/live/customize.rs:29]: Let's put these in the first block with the other crate imports.

File: src/live/customize.rs
Comment (bgilbert) [inline src/live/customize.rs:182]: We generally avoid bare `?` operators when calling external libraries.  Let's add `context()` or `with_context()` calls that explain what we were doing at the time of the failure.

File: src/live/customize.rs
Comment (bgilbert) [reply src/live/customize.rs:184]: Yup, it [looks like](https://github.com/nmstate/nmstate/blob/base/rust/src/lib/error.rs) `NmstateError` doesn't implement `Error`.  Ideally it would do so, but if necessary you can wrap the error in `anyhow!()` here.

File: src/live/customize.rs
Comment (bgilbert) [inline src/live/customize.rs:188]: Hmm, won't this create something like `my-config.yaml.nmconnection`?

File: src/live/customize.rs
Comment (bgilbert) [inline src/live/customize.rs:189]: Looks like we lost the overwrite check that `network_keyfile()` performs here.

**Review by bgilbert — COMMENTED** [review]
This is great, thank you!

File: src/cmdline/mod.rs
Comment (qinqon) [reply src/cmdline/mod.rs:284]: Can we enforce this with `cargo fmt` ?

Comment (bgilbert) [reply src/cmdline/mod.rs:284]: The issue is that the short help text needs to be a single line of a certain maximum length, which varies depending on the width of the flags.  I'm not sure `cargo fmt` is configurable to that degree.

File: src/live/customize.rs
Comment (qinqon) [reply src/live/customize.rs:188]: The final PR will just use base name so the extension is skip and we get `my-config.nmconnection`.

File: src/cmdline/mod.rs
Comment (qinqon) [reply src/cmdline/mod.rs]: Done

File: src/cmdline/mod.rs
Comment (qinqon) [reply src/cmdline/mod.rs]: Done

File: src/live/customize.rs
Comment (qinqon) [reply src/live/customize.rs:184]: Going to try to make the right thing and put a PR at Nmstate so we have the proper NmstateError type there.

File: src/cmdline/mod.rs
Comment (qinqon) [reply src/cmdline/mod.rs]: Done

File: src/cmdline/mod.rs
Comment (qinqon) [reply src/cmdline/mod.rs]: Done

File: src/live/customize.rs
Comment (qinqon) [reply src/live/customize.rs]: Done

File: src/live/customize.rs
Comment (qinqon) [reply src/live/customize.rs]: Done

File: src/live/customize.rs
Comment (qinqon) [reply src/live/customize.rs]: Done

File: src/live/customize.rs
Comment (qinqon) [reply src/live/customize.rs]: Created PR at nmstate implementing the trait https://github.com/nmstate/nmstate/pull/1912

**Comment by qinqon** [comment]
Testing part is not fixed yet, but code is reviewable now.

File: tests/images/customize.sh
Comment (qinqon) [reply tests/images/customize.sh:61]: After changing them to add the condition for generate NetworkManager keyfile I have try to run the tests with 
```bash
./tests/images.sh ~/Documents/cnv/sandbox/coreos-live/
```
And the output is the following, looks like the checksums does not match.
```bash
Boot media will automatically install to /dev/vda without confirmation.
1189: digest inplace-iso
222: sha256sum inplace-iso
222: awk '{print $1}'
189: '[' e41247f55606742bd0083c20ffc53bd92b8001d015165614f5858c37efb7b015 = fb49271eda68f2fa7ce2c4727411e3bfd30f099a93a3784d7469d4dedd63e54d ']'
1: rm -rf /tmp/tmp.sNbyWtpoKh
```

Do you know what's going on here ?

File: src/cmdline/mod.rs
Comment (qinqon) [reply src/cmdline/mod.rs:284]: I would say `nmstate_file` since it's important to state that it's a file to the state not the state embedded at the argument.

File: src/live/customize.rs
Comment (qinqon) [reply src/live/customize.rs:193]: Adding "context" is not enough to have a proper return that we can pin point ?

File: src/live/customize.rs
Comment (qinqon) [reply src/live/customize.rs:195]: Converting the "file_stem" to str.

File: src/cmdline/mod.rs
Comment (qinqon) [reply src/cmdline/mod.rs]: Done

File: src/cmdline/mod.rs
Comment (qinqon) [reply src/cmdline/mod.rs:284]: Done

File: src/live/customize.rs
Comment (qinqon) [reply src/live/customize.rs]: Done

File: tests/images/customize.sh
Comment (qinqon) [reply tests/images/customize.sh]: Done

File: tests/images/customize.sh
Comment (qinqon) [reply tests/images/customize.sh:61]: I am going to proposa a PR a change some we can instruct it to generate the keyfile but without the UUID.

File: src/live/customize.rs
Comment (cgwalters) [reply src/live/customize.rs:196]: You can avoid the `to_str()` and `unwrap()` here by using [camino](https://crates.io/crates/camino) - we use it extensively in the ostree stack which hard requires UTF-8 pathnames.

Comment (qinqon) [reply src/live/customize.rs:193]: Done

Comment (bgilbert) [reply src/live/customize.rs:193]: @cgwalters Let's not add an additional dependency just to handle one call site.  Pathname conversion is a recurring pattern throughout the codebase.

@qinqon Apologies for the noise, but please revert the camino addition.  If we want to change the way we handle pathnames, we should do it comprehensively as a separate PR.

As to your question, adding `context` is the correct approach, but then the `unwrap()` panics if the conversion fails.  Please use `?` to return an error instead.

File: src/cmdline/mod.rs
Comment (bgilbert) [reply src/cmdline/mod.rs:284]: I was suggesting that we use consistent terminology in the help text.  We're currently saying "Nmstate file" in the short help and "Nmstate config" in the long help.

As to the argument name, I think the original `--network-nmstate` (or maybe `--network-nmstate-file` if it fits) is clearer than the updated `--nmstate-file`, since the first one helpfully groups the network configuration arguments together.  However, I'd be interested in other people's thoughts here.

File: src/live/customize.rs
Comment (qinqon) [reply src/live/customize.rs]: No problem, at least I have learn about `camino` now, reverted and removed unwrap.

File: src/cmdline/mod.rs
Comment (qinqon) [reply src/cmdline/mod.rs:284]: What about removing `nmstate` and keeping `--network-config-file` ? is similar to how it's done at openshift installer at the baremetal host [1].

[1] https://github.com/openshift/installer/blob/969378dc498336d1ac29baa89f75324a44467d5d/pkg/types/baremetal/platform.go#L43

**Comment by qinqon** [comment]
Cargo.toml has now a pin to https://github.com/nmstate/nmstate/pull/1915, test should pass, but this cannot be merged until the PR is merge and nmstate published to cargo.io.

**Comment by qinqon** [comment]
@bgilbert can you help me find why CI is failing ? I think all the bits are in place now (except the TODOs).

**Comment by bgilbert** [comment]
@qinqon `live.ign` (and `dest.ign`) is checking for the existence of `installer-test-nmstate.nmconnection`, but the tests are no longer creating that file.

**Comment by qinqon** [comment]
@bgilbert fixed `pr-merge` job, do we have to do something for `x390x` to work or this is a WIP job ?

**Comment by qinqon** [comment]
closed/re-open by mistake.

**Comment by bgilbert** [comment]
The s390x jobs broke recently and we're not sure why.  This change doesn't affect any s390x-specific code, so I think it's reasonable to ignore those jobs.

File: src/cmdline/mod.rs
Comment (qinqon) [reply src/cmdline/mod.rs:284]: @bgilbert what is the best name for this ?

**Comment by bgilbert** [comment]
FYI, the s390x jobs are working again, but you'll have to rebase to pick up the fix.

**Comment by qinqon** [comment]
@bgilbert every issue is addressed now, it's ready for review.

**Comment by qinqon** [comment]
/hold
Looks like nmstate gen_conf rust should return the keyfiles name as it was like that at python version, let's wait for that to consume it.

**Comment by bgilbert** [comment]
@qinqon Is the nmstate Rust crate packaged in Fedora?  The `RPMs` test thinks it isn't, and I can't find it in the `nmstate` subpackages either.  To keep coreos-installer shippable downstream, we generally avoid landing PRs here until their dependencies are available in Fedora.

**Comment by qinqon** [comment]
> @qinqon Is the nmstate Rust crate packaged in Fedora? The `RPMs` test thinks it isn't, and I can't find it in the `nmstate` subpackages either. To keep coreos-installer shippable downstream, we generally avoid landing PRs here until their dependencies are available in Fedora.

Not yet, they are working on it, ist just a matter of time.

**Comment by bgilbert** [comment]
For the record, this PR increases the size of the release binary without debug symbols from 12.1 MiB to 15.0 MiB.

**Comment by qinqon** [comment]
> For the record, this PR increases the size of the release binary without debug symbols from 12.1 MiB to 15.0 MiB.

Is it worth to investigate the nmstate side before merging this PR or we can do a follow up ?

**Comment by bgilbert** [comment]
I was concerned about the number of additional dependencies we were pulling in, which is why I checked, but I'm personally not too worried about 3 MiB.  (Maybe I'm just used to Go binaries.)  If there are ways to reduce the size, that'd be great, but it can be done as a followup.

**Comment by qinqon** [comment]
> I was concerned about the number of additional dependencies we were pulling in, which is why I checked, but I'm personally not too worried about 3 MiB. (Maybe I'm just used to Go binaries.) If there are ways to reduce the size, that'd be great, but it can be done as a followup.

I will prepare a follow up PR at nmstate using "features" from rust so we have "default" with "show" and "apply" and "gc" for "gen_conf" this we only stuff needed for a feature get compiled.

**Review by bgilbert — COMMENTED** [review]
This looks great!  Just some nits.

Docs need to be regenerated with `make docs`.

File: src/cmdline/mod.rs
Comment (bgilbert) [reply src/cmdline/mod.rs]: `--network-nmstate` sounds good, thanks. :+1: `--network-config-file` implies that a particular file format is the canonical one, which isn't the case here.

File: fixtures/customize/installer-test-nmstate-json-0.nmconnection
Comment (qinqon) [reply fixtures/customize/installer-test-nmstate-json-0.nmconnection:1]: We are not testing "nmstatectl gc" we are testing it's integration, but it's true that a bug at the integration can dump the wrong file, I can add a check for content too if needed.

File: Cargo.toml
Comment (bgilbert) [reply Cargo.toml]: `nix` < `nmstate` :slightly_smiling_face:

File: fixtures/customize/installer-test-nmstate-json-0.nmconnection
Comment (bgilbert) [reply fixtures/customize/installer-test-nmstate-json-0.nmconnection:1]: Yeah, I don't think it's worth a lot of effort, but maybe the test could do `coreos-installer iso network extract -C` and then grep the configs for `nmstate-eth<n>`?

Comment (qinqon) [reply fixtures/customize/installer-test-nmstate-json-0.nmconnection:1]: At the end the nmstate team is going back to adding the generated nmconection file name to the gen_conf output (it was included on the python nmstate version) so using the path also verify partially the content, I think that's good enough.

**Comment by qinqon** [comment]
@bgilbert addressed comments and adapted to latest nmstate so the keyfile name is generated by it, we have to wait for the new crate/rpm.

File: src/live/customize.rs
Comment (bgilbert) [inline src/live/customize.rs:198]:
```suggestion
                bail!("config already specifies keyfile {}", nm_con_path_str);
```

**Comment by bgilbert** [comment]
LGTM!  I think the remaining things are:

- [ ] Fix clippy lints
- [ ] nmstate 2.1.2 crate released and `Cargo.toml` updated
- [ ] nmstate Rust crate available in Fedora updates so RPMs CI is green

**Comment by qinqon** [comment]
@bgilbert fixed clippy issues.

**Comment by qinqon** [comment]
Hey @qinqon Looks like fedora 36 update is ready at https://bodhi.fedoraproject.org/updates/FEDORA-2022-b78be96657
But the coreos-install build RPM need to add requirements ` (crate(nmstate/default) >= 2.1.2 with crate(nmstate/default) < 3.0)`

**Comment by bgilbert** [comment]
Rust package requirements are automatically added by `rust-srpm-macros`.

**Comment by qinqon** [comment]
@bgilbert Do you know what happening with CI ?

**Comment by bgilbert** [comment]
The MSRV tests should be fixed by https://github.com/coreos/coreos-installer/pull/926.  The RPM test is awaiting [FEDORA-2022-b78be96657](https://bodhi.fedoraproject.org/updates/FEDORA-2022-b78be96657) going stable.

**Comment by bgilbert** [comment]
I'm not seeing that ENOSPC error in other PRs.

**Comment by bgilbert** [comment]
Could you try rebasing?  That should fix the GitHub Actions checks, and we can see whether it fixes the kola failure too.  If it doesn't, there may be an actual problem with the PR.

**Comment by jlebon** [comment]
I think @bgilbert and I came to the same realization in parallel. :) I suspect CI is hitting `ENOSPC` because `rdcore` includes debuginfo and we're not using LTO in debug mode so it's including nmstate. But even without this change, I think we should probably strip the binary when installing it in debug mode like CI does. Did that in https://github.com/coreos/coreos-installer/pull/929.

**Comment by bgilbert** [comment]
We're testing with 1 GB of RAM, and this PR increases the `rdcore` size from 132 to 178 MiB.

(Actually we're not using LTO at all anymore; it was too buggy.)

**Comment by bgilbert** [comment]
Confirmed that https://github.com/coreos/coreos-installer/pull/929 fixes the `ENOSPC`.

**Comment by qinqon** [comment]
> We're testing with 1 GB of RAM, and this PR increases the `rdcore` size from 132 to 178 MiB.
> 
> (Actually we're not using LTO at all anymore; it was too buggy.)

Do we have to do something at nmstate to reduce `rdcore` size ? or 178 is ok ?

**Comment by bgilbert** [comment]
No, the rdcore binary is only about 2 MiB with all symbols stripped, or 6 MiB with debug symbols stripped.  That's what we actually ship in the OS, so this is only a CI problem.  We'll just land #929 and then you can rebase on top of it.

**Comment by bgilbert** [comment]
Okay, if you rebase this on `main` it should fix CI.

**Review by bgilbert — APPROVED** [review]
Thanks for all your work on this! :tada: :balloon: :fireworks:

**Pull request merged**
